language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/CommandsSubmission.java

```diff
@@ -0,0 +1,246 @@
+ * for initial creation and extra builder methods for additional parameters
+ * e.g {@link #withActAs(List)}, {@link #withWorkflowId(String)} etc.
+ *
+ * Usage:
+ * <pre>
+ *   var params = CommandsSubmission.create("applicationId", "commandId", commands)
+ *                                   .withAccessToken("token")
+ *                                   .withParty("party")
+ *                                   .with...
+ * <pre/>
+ */
+public final class CommandsSubmission {
+  private String applicationId;
+  private String commandId;
+  private List<@NonNull ? extends HasCommands> commands;
+
+  private Optional<String> workflowId;
+  private List<@NonNull String> actAs;
+  private List<@NonNull String> readAs;
+  private Optional<Instant> minLedgerTimeAbs;
+  private Optional<Duration> minLedgerTimeRel;
+  private Optional<Duration> deduplicationTime;
+  private Optional<String> accessToken;
+
+  private CommandsSubmission(
+      Optional<String> workflowId,
+      String applicationId,
+      String commandId,
+      List<@NonNull String> actAs,
+      List<@NonNull String> readAs,
+      Optional<Instant> minLedgerTimeAbs,
+      Optional<Duration> minLedgerTimeRel,
+      Optional<Duration> deduplicationTime,
+      List<@NonNull ? extends HasCommands> commands,
+      Optional<String> accessToken) {
+    this.workflowId = workflowId;
+    this.applicationId = applicationId;
+    this.commandId = commandId;
+    this.actAs = actAs;
+    this.readAs = readAs;
+    this.minLedgerTimeAbs = minLedgerTimeAbs;
+    this.minLedgerTimeRel = minLedgerTimeRel;
+    this.deduplicationTime = deduplicationTime;
+    this.commands = commands;
+    this.accessToken = accessToken;
+  }
+
+  public static CommandsSubmission create(
+      String applicationId, String commandId, List<@NonNull ? extends HasCommands> commands) {
+    return new CommandsSubmission(
+        Optional.empty(),
+        applicationId,
+        commandId,
+        emptyList(),
+        emptyList(),
+        empty(),
+        empty(),
+        empty(),
+        commands,
+        empty());
+  }
+
+  public Optional<String> getWorkflowId() {
+    return workflowId;
+  }
+
+  public String getApplicationId() {
+    return applicationId;
+  }
+
+  public String getCommandId() {
+    return commandId;
+  }
+
+  public List<String> getActAs() {
+    return unmodifiableList(actAs);
+  }
+
+  public List<String> getReadAs() {
+    return unmodifiableList(readAs);
+  }
+
+  public Optional<Instant> getMinLedgerTimeAbs() {
+    return minLedgerTimeAbs;
+  }
+
+  public Optional<Duration> getMinLedgerTimeRel() {
+    return minLedgerTimeRel;
+  }
+
+  public Optional<Duration> getDeduplicationTime() {
+    return deduplicationTime;
+  }
+
+  public List<? extends HasCommands> getCommands() {
+    return unmodifiableList(commands);
+  }
+
+  public Optional<String> getAccessToken() {
+    return accessToken;
+  }
+
+  public CommandsSubmission withWorkflowId(String workflowId) {
```

Comment:
All this duplication, points to a future of implementing Java clients in Scala instead ;-)